File: app/src/main/java/com/knotworking/numbers/MainActivity.java
package com.knotworking.numbers;

import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;

import com.knotworking.numbers.counter.CreateItemDialog;
import com.knotworking.numbers.database.DatabaseHelper;
import com.knotworking.numbers.database.DatabaseHelperImpl;

import static com.knotworking.numbers.Constants.CONVERTER_TAB;
import static com.knotworking.numbers.Constants.COUNTER_TAB;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    private static final String TAG = MainActivity.class.getSimpleName();
    private static final String NEW_ITEM_DIALOG = "new_item_dialog";

    private NumbersPagerAdapter adapter;
    private ViewPager pager;
    private FloatingActionButton fab;
    private DatabaseHelper databaseHelper;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        databaseHelper = new DatabaseHelperImpl(this);
        adapter = new NumbersPagerAdapter(getSupportFragmentManager());
        pager = (ViewPager)findViewById(R.id.activity_view_pager);
        pager.setAdapter(adapter);
        fab = (FloatingActionButton)findViewById(R.id.activity_fab);
        fab.setOnClickListener(this);

        setupActionBar();

    }

    private void setupActionBar() {
        final ActionBar actionBar = getSupportActionBar();
        actionBar.setNavigationMode(ActionBar.NAVIGATION_MODE_TABS);
        actionBar.setDisplayShowTitleEnabled(false);

        ActionBar.TabListener tabListener = new ActionBar.TabListener() {
            @Override
            public void onTabSelected(ActionBar.Tab tab, FragmentTransaction ft) {
                pager.setCurrentItem(tab.getPosition());
                if (tab.getPosition() == COUNTER_TAB) {
                    fab.show();
                } else {
                    fab.hide();
                }
            }

            @Override
            public void onTabUnselected(ActionBar.Tab tab, FragmentTransaction ft) {

            }

            @Override
            public void onTabReselected(ActionBar.Tab tab, FragmentTransaction ft) {

            }
        };

        for (int i = 0; i < 2; i++) {
            actionBar.addTab(actionBar.newTab()
            .setText(getTabTitle(i))
            .setTabListener(tabListener));
        }

        pager.setOnPageChangeListener(new ViewPager.SimpleOnPageChangeListener() {
            @Override
            public void onPageSelected(int position) {
                actionBar.setSelectedNavigationItem(position);
            }
        });
    }

    private int getTabTitle(int position) {
        switch (position) {
            case CONVERTER_TAB:
                return R.string.first_tab;
            case COUNTER_TAB:
                return R.string.second_tab;
            default:
                Log.e(TAG, "tab position not recognised");
                return 0;
        }
    }

    @Override
    public void onClick(View view) {
        if (view.getId() == R.id.activity_fab) {
            showNewItemDialog();
        }
    }

    private void showNewItemDialog() {
        FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
        Fragment prev = getSupportFragmentManager().findFragmentByTag(NEW_ITEM_DIALOG);
        if (prev != null) {
            ft.remove(prev);
        }
        ft.addToBackStack(null);

        CreateItemDialog dialog = new CreateItemDialog();
        dialog.show(ft, NEW_ITEM_DIALOG);
    }
}

File: app/src/main/java/com/knotworking/numbers/MainActivity.kt
package com.knotworking.numbers

import android.os.Bundle
import android.preference.PreferenceManager
import android.support.design.widget.FloatingActionButton
import android.support.v4.app.FragmentTransaction
import android.support.v4.view.ViewPager
import android.support.v7.app.ActionBar
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View
import com.knotworking.numbers.Constants.CONVERTER_TAB
import com.knotworking.numbers.Constants.COUNTER_TAB
import com.knotworking.numbers.Constants.EXCHANGE_RATE_FETCH_TIME
import com.knotworking.numbers.api.CurrencyApi
import com.knotworking.numbers.counter.CreateItemDialog
import java.util.concurrent.TimeUnit

class MainActivity : AppCompatActivity(), View.OnClickListener {

    private var adapter: NumbersPagerAdapter? = null
    private var pager: ViewPager? = null
    private var fab: FloatingActionButton? = null

    //TODO inject singleton
    private val currencyApi = CurrencyApi(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        adapter = NumbersPagerAdapter(supportFragmentManager)
        pager = findViewById<View>(R.id.activity_view_pager) as ViewPager
        pager?.adapter = adapter
        fab = findViewById<View>(R.id.activity_fab) as FloatingActionButton
        fab?.setOnClickListener(this)

        setupActionBar()
        checkExchangeRates()
    }

    private fun setupActionBar() {
        val actionBar = supportActionBar
        actionBar?.navigationMode = ActionBar.NAVIGATION_MODE_TABS
        actionBar?.setDisplayShowTitleEnabled(false)

        val tabListener = object : ActionBar.TabListener {
            override fun onTabSelected(tab: ActionBar.Tab, ft: FragmentTransaction) {
                pager?.currentItem = tab.position
                if (tab.position == COUNTER_TAB) {
                    fab?.show()
                } else {
                    fab?.hide()
                }
            }

            override fun onTabUnselected(tab: ActionBar.Tab, ft: FragmentTransaction) {

            }

            override fun onTabReselected(tab: ActionBar.Tab, ft: FragmentTransaction) {

            }
        }

        for (i in 0..1) {
            actionBar?.addTab(actionBar.newTab()
                    .setText(getTabTitle(i))
                    .setTabListener(tabListener))
        }

        pager!!.setOnPageChangeListener(object : ViewPager.SimpleOnPageChangeListener() {
            override fun onPageSelected(position: Int) {
                actionBar?.setSelectedNavigationItem(position)
            }
        })
    }

    private fun getTabTitle(position: Int): Int {
        when (position) {
            CONVERTER_TAB -> return R.string.first_tab
            COUNTER_TAB -> return R.string.second_tab
            else -> {
                Log.e(TAG, "tab position not recognised")
                return 0
            }
        }
    }

    override fun onClick(view: View) {
        if (view.id == R.id.activity_fab) {
            showNewItemDialog()
        }
    }

    private fun showNewItemDialog() {
        val ft = supportFragmentManager.beginTransaction()
        val prev = supportFragmentManager.findFragmentByTag(NEW_ITEM_DIALOG)
        if (prev != null) {
            ft.remove(prev)
        }
        ft.addToBackStack(null)

        val dialog = CreateItemDialog()
        dialog.show(ft, NEW_ITEM_DIALOG)
    }

    private fun checkExchangeRates() {

        //TODO areExchangeRatesInDb

        //TODO refactor into new method
        val fetchThreshold = TimeUnit.DAYS.toMillis(7)
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)
        val lastFetchTimeStamp = sharedPreferences.getLong(EXCHANGE_RATE_FETCH_TIME, 0)
        val now = System.currentTimeMillis()
        val diff = now - lastFetchTimeStamp

        if (diff > fetchThreshold) {
            currencyApi.getExchangeRates()
        }
    }

    companion object {
        private val TAG = MainActivity::class.java.simpleName
        private val NEW_ITEM_DIALOG = "new_item_dialog"
    }
}
